Neos.Neos/Classes/FrontendRouting/Options.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\FrontendRouting;
+
+/**
+ * Immutable filter DTO for {@see NodeUriBuilder::uriFor()}
+ *
+ * Example:
+ *
+ *   Options::create(forceAbsolute: true);
+ *
+ * @api for the factory methods; NOT for the inner state.
+ */
+final readonly class Options
+{
+    /**
+     * @internal the properties themselves are readonly; only the write-methods are API.
+     * @param array<string, mixed> $routingArguments
+     */
+    private function __construct(
+        public bool $forceAbsolute,
+        public string $format,
+        public array $routingArguments,
+    ) {
+    }
+
+    /**
+     * Creates an instance with the specified options
+     *
+     * Note: The signature of this method might be extended in the future, so it should always be used with named arguments
+     * @see https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments
+     */
+    public static function create(
+        bool $forceAbsolute = null
+    ): self {
+        return new self(
+            $forceAbsolute ?? false,
+            '',
+            []
+        );
+    }
+
+    /**
+     * Returns a new instance with the specified additional options
+     *
+     * Note: The signature of this method might be extended in the future, so it should always be used with named arguments
+     * @see https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments
+     */
+    public function with(
+        bool $forceAbsolute = null
+    ): self {
+        return new self(
+            $forceAbsolute ?? $this->forceAbsolute,
+            $this->format,
+            $this->routingArguments
+        );
+    }
+
+    /**
+     * Option to set a custom routing format
+     *
+     * In order for the routing framework to match and resolve this format,
+     * your have to define a custom route in Routes.yaml
+     *
+     *   -
+     *     name: 'Neos :: Frontend :: Document node with json format'
+     *     uriPattern: '{node}.json'
+     *     defaults:
+     *       '@package': Neos.Neos
+     *       '@controller': Frontend\Node
+     *       '@action': show
+     *       '@format': json
+     *     routeParts:
+     *       node:
+     *         handler: Neos\Neos\Routing\FrontendNodeRoutePartHandlerInterface
+     *
+     * See also {@link https://docs.neos.io/guide/rendering/rendering-special-formats}
+     */
+    public function withCustomFormat(string $format): self
```

Comment:
i also noticed that the current format will NOT be takend from the `actionRequest` so one would need to take care of this themselves is this ok?

Reply:
But i guess the Fusion helpers should default to the requests format? Either in php or better via

```
format = ${request.format}
```

what do you think?

Reply:
Good question, I'm not sure about that. When I played with HTMX, I used a custom format for the HX requests (see https://github.com/bwaidelich/Wwwision.HtmxTest/blob/main/Resources%2FPrivate%2FFusion%2FComponents%2FHtmxUri.fusion)
In the results I would _not_ want this format to be used for all links implicitly

Reply:
Currently it's the expected behavior, I have had it in both directions in projects. Sometimes I was happy it would automatically keep the xml I had in hte current request without me having to check it at every link, other times I would have wished to ignore it and use html by default 🤷

Reply:
yeah i guess for fusion (especially as the magic is currently still part of the `Neos.Fusion:ActionUri`) i think it would make sense to keep it for the `Neos.Neos:NodeUri` as well.

Ill vote for making it a default value on fusion side so its directly visible to the human eye? or should it be part of the fusion php implementation?

Reply:
My gut feeling is still: html is like the 99% case and I would want to explicitly change that if I wanted to create links to other resources in the same format (e.g. in a JSON response I _might_ want to link to other JSON resources, but I would expect the node URI to still have the default format:
```json
{
  "title": "foo",
  "contactUri": "https://domain.tld/contact",
  "includedResources": {
    "contacts": [
      {"id": "123", "href": "/api/contacts/123.json"}
    ]
  }
}
```

Or maybe I can override the default format for the `RoutingDefaults` per controller/action?

Reply:
I agree that the magic of keeping the format either has helped (by knowing it beforehand) or worse lead to confusion and i had to debug many times (because i forgot) why uri building failed ... the `@format: json` part in the exception dump `cat Data...` tends to go unnoticed for the untrained eye.
Still for this change it would be cruel to change behaviour like this ever so slightly and id prefer a dedicated discussion and separate pr.
I think we only have to agree that the PHP Node Uri building api doesnt do it by default and how to implement this magic again for the Fusion prototypes.

Reply:
Im not against removing the format by request magic but id prefer a separate pr for that (and probably wont do it myself right now because i cant sell it as well as you do :D)
I guess this pr is a step in the right direction by removing that feature from the core node uribuilding and making it opt in from the outside. That means we can eventually get rid of it!